### DISCUSSION TITLE
Need help with direct register manipulation

### DISCUSSION THREAD

**Discussion by imeshsps** [CommunityGD32Cores/GD32Core-New#75] [Q&A] [answered]
I was trying to get free watchdog timer working by writing directly to registers and by writing the following I was able to start the watchdog timer but it keeps restarting. I'm trying to configure it but so far no luck.

 `FWDGT_CTL = 0X5555;
  FWDGT_RLD = FWDGT_RLD | 0x00000800;
  FWDGT_CTL = 0xCCCC;`
  
and this line to reset the watchdog timer according to the datasheet.

`FWDGT_CTL = 0xAAAA;`

with these, I can compile it on Arduino IDE.

I was also trying to do some GPIO control the same way but when I try to write something to GPIOC_CTL it gives the following error when trying to compile.

`In function 'void setup()':
Register_access:11:3: error: 'GPIOC_CTL' was not declared in this scope; did you mean 'GPIO_CTL'?
   11 |   GPIOC_CTL = GPIOC_CTL | 0x08000000;
      |   ^~~~~~~~~
      |   GPIO_CTL
exit status 1
'GPIOC_CTL' was not declared in this scope; did you mean 'GPIO_CTL'?
`

Is it using a different naming convention? Any help is really appreciated. Thanks.

**Comment by maxgerhardt** [comment, marked as answer]
See 

https://github.com/CommunityGD32Cores/ArduinoCore-GD32/blob/a22a2e3aca8ce338f1ca8e2170fd70c53399120e/system/GD32F1x0_firmware/GD32F1x0_standard_peripheral/Include/gd32f1x0_gpio.h#L45-L54

So you would have to do

```cpp
GPIO_CTL(GPIOC) |= 0x08000000;
```

**Reply by imeshsps** [reply]
![11](https://user-images.githubusercontent.com/42859952/184125716-fb9edd94-d3d7-4027-8624-21b398c2a485.png)
![12](https://user-images.githubusercontent.com/42859952/184125744-9107cf1a-1f9e-4adf-a82a-613f69c320af.png)
Thank you for your reply and pointing me to that file. I was going through files looking for this for like days now. 

` GPIO_CTL(GPIOC) |= (B1 << 26);`
  `GPIO_OCTL(GPIOC) &= ~(B1 << 13);`

These two lines should turn PC13 to an output and output logic LOW right? Anything I'm missing? I'm having a hard time trying to turn on inbuilt led. The blink example is working fine on this board. Any help is really appreciated. 

Sorry for asking so many questions. I'm new to this register level working and trying to get some understanding. Thanks.

**Reply by maxgerhardt** [reply]
I would write just `1u` instead of `B1`. Otherwise this looks good.

However, did you enable the GPIO clock for GPIOC beforehand? Otherwise all changes are ignored.
```cpp
rcu_periph_clock_enable(RCU_GPIOC);
```

**Reply by imeshsps** [reply]
> However, did you enable the GPIO clock for GPIOC beforehand? Otherwise all changes are ignored.

I did not knew about that. It is working now. Thank you very much. 
You just saved me from hours of head scratching. 

Can you point me to any material that can help me to get a better understanding? Thanks in advance.

**Reply by maxgerhardt** [reply]
I would always take a look at what the **SPL** examples do, because they're more low-level. See e.g. https://github.com/CommunityGD32Cores/gd32-pio-projects/blob/main/gd32-spl-blinky/src/main.c, or straight from the source https://github.com/CommunityGD32Cores/gigadevice-firmware-and-docs/tree/main/GD32F1x0/GD32F1x0_Firmware_Library_v3.1.0/Examples. Of course for the ultra low level stuff, you have to look into the SPL implementation just like I've linked above, but usually they do very little work there and go straight to the registers.

**Reply by maxgerhardt** [reply]
If you are using PlatformIO and an SWD-capable probe (like an ST-Link), you can also leverage the power of debugging as explained in https://github.com/CommunityGD32Cores/gd32-pio-projects#debugging. In the debug view, due to having the SVD loaded, you can also take a look at each hardware peripheral (like GPIOC, or RCU) and step over your code line by line and see the registers change. 

If you e.g. had done this project in PlatformIO and debugged it, you would have seen that changes to the GPIOC registers are ignored although your code writes to it, and so you may have gotten the idea that you have to do "something" so that the GPIOC peripheral register reacts at all, which is, activating the clock for the peripheral. (Btw, this goes for any peripheral, they won't do much if their clock is not enabled in the RCU).

I would not wish writing low-level code with no debugger in the standard Arduino IDE with no debugger to anyone. Always leverage the best tools you can.

**Reply by imeshsps** [reply]
Thank you very much for your guidance. This should help me to get started. I'm using the ST-Link but currently programming with Arduino IDE. I have installed PlatformIO in VScode but I don't have much experience with that. I'll try to get more familiar with that. Thank you very much for your support again.

**Reply by maxgerhardt** [reply]
And finally, like the STM32 core, we have macros that give you the right registers for port manipulation. 

https://github.com/CommunityGD32Cores/ArduinoCore-GD32/blob/a22a2e3aca8ce338f1ca8e2170fd70c53399120e/cores/arduino/gd32/pins_arduino.h#L184-L204

https://github.com/CommunityGD32Cores/ArduinoCore-GD32/blob/f6dc080e2524877910f4b8c6240bfbb7901aab44/cores/arduino/gd32/pins_arduino.h#L218-L224

So you can write code like

```cpp
auto myPin = PC13;
auto pinMask = digitalPinToBitMask(myPin); // (1 << 13)
auto port = digitalPinToPort(myPin); // GPIOC
auto outputReg = portOutputRegister(port); // GPIOC OCTL

*outputReg |= pinMask;
delay(1000);
*outputReg &= ~pinMask;
```

Adafaruit libraries like the https://github.com/adafruit/Adafruit_BusIO) also expect these macros, and they work (tested with the OLED SSD1306 library).

I **heavily** recommend doing port manipulations this way, since this is the standardized Arduino way.
